// unjwt@0.7.1/dist/_chunks/jws.mjs
import { C as validateCriticalHeadersJWS, D as JWTError, b as getPlaintextBytes, f as applyTypCtyDefaults, h as computeJwtTimeClaims, o as isJWK, s as isJWKSet } from "./utils.mjs";
import { a as base64UrlEncode, i as base64UrlDecode, o as textDecoder, s as textEncoder, u as sanitizeObjectCopy } from "./libs/unsecure.mjs";
import { T as verify$1, h as importKey, p as getJWKsFromSet, w as sign$1 } from "./jwk.mjs";
import { a as decodeJWSPayload, c as resolveSigningKey, i as checkAlgAllowed, l as validateJwtClaimsIfJsonPayload, n as JWS_ALG_CTX, o as decodeProtectedHeader, r as buildJWKSetFilter } from "./_internal.mjs";
async function signMulti(payload, signers, options = {}) {
	if (!Array.isArray(signers) || signers.length === 0) throw new JWTError("signMulti requires at least one signer.", "ERR_JWS_INVALID");
	const resolvedAlgs = signers.map((s, i) => _resolveSignerAlg(s.key, i));
	const b64 = _resolveB64(signers);
	const payloadBytes = getPlaintextBytes(computeJwtTimeClaims(payload, options) || payload);
	const payloadEncoded = b64 ? base64UrlEncode(payloadBytes) : textDecoder.decode(payloadBytes);
	const wireSignatures = [];
	for (let i = 0; i < signers.length; i++) {
		const signer = signers[i];
		const alg = resolvedAlgs[i];
		const protectedHeader = _buildSignerProtectedHeader(signer, alg, payload, b64);
		const unprotectedHeader = sanitizeObjectCopy(signer.unprotectedHeader);
		_assertDisjointHeaders(protectedHeader, unprotectedHeader);
		const protectedHeaderEncoded = base64UrlEncode(JSON.stringify(protectedHeader));
		const signingInputBytes = textEncoder.encode(`${protectedHeaderEncoded}.${payloadEncoded}`);
		const wireSig = {
			protected: protectedHeaderEncoded,
			signature: base64UrlEncode(await sign$1(alg, await resolveSigningKey(alg, await importKey(signer.key, {
				alg,
				expect: "private"
			}), "sign"), signingInputBytes))
		};
		if (unprotectedHeader && Object.keys(unprotectedHeader).length > 0) wireSig.header = unprotectedHeader;
		wireSignatures.push(wireSig);
	}
	return {
		payload: payloadEncoded,
		signatures: wireSignatures
	};
}
async function verifyMulti(jws, keyOrLookup, options = {}) {
	const general = _normalizeSerialization(jws);
	if (typeof general.payload !== "string" || !Array.isArray(general.signatures) || general.signatures.length === 0) throw new JWTError("Invalid JWS JSON Serialization: missing payload or signatures.", "ERR_JWS_INVALID_SERIALIZATION");
	let lastError;
	let cryptoAttempted = false;
	for (let i = 0; i < general.signatures.length; i++) {
		const wireSig = general.signatures[i];
		if (typeof wireSig.signature !== "string") {
			lastError = new JWTError(`Signature[${i}] is missing the "signature" field.`, "ERR_JWS_INVALID");
			continue;
		}
		let protectedHeader;
		try {
			protectedHeader = decodeProtectedHeader(wireSig.protected, "JWS");
		} catch (err) {
			lastError = err instanceof JWTError ? err : void 0;
			continue;
		}
		if (!protectedHeader.alg) {
			lastError = new JWTError(`Invalid JWS: Signature[${i}] protected header must have an "alg" property.`, "ERR_JWS_INVALID");
			continue;
		}
		const unprotectedHeader = sanitizeObjectCopy(wireSig.header);
		_assertDisjointHeaders(protectedHeader, unprotectedHeader);
		const effective = _mergeHeaders(protectedHeader, unprotectedHeader);
		const alg = protectedHeader.alg;
		if (options.typ && protectedHeader.typ !== options.typ) {
			lastError = new JWTError(`Invalid JWS: "typ" Header Parameter mismatch on signature[${i}]. Expected "${options.typ}", got "${protectedHeader.typ}".`, "ERR_JWS_INVALID");
			continue;
		}
		const rawKeyMaterial = typeof keyOrLookup === "function" ? await keyOrLookup(effective, wireSig.protected ?? "") : keyOrLookup;
		const algError = checkAlgAllowed(alg, rawKeyMaterial, options.algorithms, JWS_ALG_CTX);
		if (algError) {
			lastError = algError;
			continue;
		}
		if (options.strictSignerMatch && !_signerKeyMatches(effective, rawKeyMaterial)) continue;
		const signatureBytes = base64UrlDecode(wireSig.signature, { returnAs: "uint8array" });
		const signingInputBytes = textEncoder.encode(`${wireSig.protected ?? ""}.${general.payload}`);
		cryptoAttempted = true;
		try {
			if (!await _verifyWithKey(alg, rawKeyMaterial, signatureBytes, signingInputBytes, effective.kid)) {
				lastError = new JWTError("JWS signature verification failed.", "ERR_JWS_SIGNATURE_INVALID");
				continue;
			}
			const useB64 = protectedHeader.b64 !== false;
			const payload = decodeJWSPayload(general.payload, useB64, options.forceUint8Array);
			validateCriticalHeadersJWS(protectedHeader, options.recognizedHeaders);
			validateJwtClaimsIfJsonPayload(payload, options);
			const result = {
				payload,
				protectedHeader,
				signerIndex: i
			};
			if (unprotectedHeader) result.signerHeader = unprotectedHeader;
			return result;
		} catch (err) {
			if (err instanceof JWTError && err.code !== "ERR_JWS_SIGNATURE_INVALID" && err.code !== "ERR_JWK_KEY_NOT_FOUND") throw err;
			lastError = err instanceof JWTError ? err : void 0;
		}
	}
	if (options.strictSignerMatch && !cryptoAttempted) throw new JWTError("No signature matched the provided key under strictSignerMatch.", "ERR_JWS_NO_MATCHING_SIGNER");
	throw lastError ?? new JWTError("JWS signature verification failed.", "ERR_JWS_SIGNATURE_INVALID");
}
async function verifyMultiAll(jws, keyResolver, options = {}) {
	const general = _normalizeSerialization(jws);
	if (typeof general.payload !== "string" || !Array.isArray(general.signatures) || general.signatures.length === 0) throw new JWTError("Invalid JWS JSON Serialization: missing payload or signatures.", "ERR_JWS_INVALID_SERIALIZATION");
	const outcomes = [];
	for (let i = 0; i < general.signatures.length; i++) outcomes.push(await _verifyOneSignature(general.payload, general.signatures[i], i, keyResolver, options));
	return outcomes;
}
function generalToFlattenedJWS(jws) {
	if (jws.signatures.length !== 1) throw new JWTError(`Flattened JWS Serialization requires exactly one signature, got ${jws.signatures.length}.`, "ERR_JWS_INVALID_SERIALIZATION");
	const [signature] = jws.signatures;
	const flattened = {
		payload: jws.payload,
		signature: signature.signature
	};
	if (signature.protected !== void 0) flattened.protected = signature.protected;
	if (signature.header !== void 0) flattened.header = signature.header;
	return flattened;
}
function _resolveSignerAlg(key, index) {
	if (!isJWK(key)) throw new JWTError(`Signer[${index}] key is not a JWK.`, "ERR_JWS_SIGNER_ALG_INFERENCE");
	const alg = key.alg;
	if (!alg) throw new JWTError(`Cannot infer "alg" for signer[${index}]: the JWK has no "alg" property.`, "ERR_JWS_SIGNER_ALG_INFERENCE");
	if (alg === "none") throw new JWTError("\"none\" is not a valid signing algorithm", "ERR_JWS_ALG_NOT_ALLOWED");
	return alg;
}
function _resolveB64(signers) {
	let decided;
	for (let i = 0; i < signers.length; i++) {
		const raw = signers[i].protectedHeader?.b64;
		if (raw === void 0) continue;
		if (decided === void 0) decided = raw;
		else if (decided !== raw) throw new JWTError(`Inconsistent "b64" across signers (RFC 7797 §3 requires all signers use the same b64 value).`, "ERR_JWS_B64_INCONSISTENT");
	}
	return decided ?? true;
}
function _buildSignerProtectedHeader(signer, alg, payload, b64) {
	const userHeader = sanitizeObjectCopy(signer.protectedHeader);
	const header = {};
	if (signer.key.kid) header.kid = signer.key.kid;
	if (userHeader) Object.assign(header, userHeader);
	header.alg = alg;
	applyTypCtyDefaults(header, payload);
	if (!b64) header.b64 = false;
	else delete header.b64;
	return header;
}
function _assertDisjointHeaders(protectedHeader, unprotectedHeader) {
	if (!protectedHeader || !unprotectedHeader) return;
	for (const k of Object.keys(unprotectedHeader)) if (k in protectedHeader) throw new JWTError(`JWS header parameter "${k}" appears in both protected and unprotected headers of a signature; they must be disjoint (RFC 7515 §7.2.1).`, "ERR_JWS_HEADER_PARAMS_NOT_DISJOINT");
}
function _mergeHeaders(protectedHeader, unprotectedHeader) {
	if (!unprotectedHeader) return protectedHeader;
	return {
		...protectedHeader,
		...unprotectedHeader
	};
}
function _normalizeSerialization(input) {
	if (!input || typeof input !== "object") throw new JWTError("Invalid JWS JSON Serialization: expected an object.", "ERR_JWS_INVALID_SERIALIZATION");
	const obj = input;
	if (Array.isArray(obj.signatures)) return obj;
	if ("signature" in obj) {
		const flattened = obj;
		const signature = { signature: flattened.signature };
		if (flattened.protected !== void 0) signature.protected = flattened.protected;
		if (flattened.header !== void 0) signature.header = flattened.header;
		return {
			payload: flattened.payload,
			signatures: [signature]
		};
	}
	throw new JWTError("Invalid JWS JSON Serialization: missing both signatures[] and flattened signature.", "ERR_JWS_INVALID_SERIALIZATION");
}
function _signerKeyMatches(effectiveHeader, keyMaterial) {
	const headerKid = effectiveHeader.kid;
	if (isJWKSet(keyMaterial)) return keyMaterial.keys.some((k) => !headerKid || k.kid === headerKid);
	if (isJWK(keyMaterial)) {
		if (headerKid && keyMaterial.kid && headerKid !== keyMaterial.kid) return false;
		return true;
	}
	return !headerKid;
}
async function _verifyWithKey(alg, rawKeyMaterial, signatureBytes, signingInputBytes, headerKid) {
	if (isJWKSet(rawKeyMaterial)) {
		const candidates = getJWKsFromSet(rawKeyMaterial, buildJWKSetFilter({
			kid: headerKid,
			alg
		}));
		if (candidates.length === 0) throw new JWTError("No key found in JWK Set.", "ERR_JWK_KEY_NOT_FOUND");
		for (const candidate of candidates) if (await verify$1(alg, await resolveSigningKey(alg, await importKey(candidate, {
			alg,
			expect: "public"
		}), "verify"), signatureBytes, signingInputBytes)) return true;
		return false;
	}
	return verify$1(alg, await resolveSigningKey(alg, await importKey(rawKeyMaterial, {
		alg,
		expect: "public"
	}), "verify"), signatureBytes, signingInputBytes);
}
async function _verifyOneSignature(payload, wireSig, index, keyResolver, options) {
	const fail = (error, protectedHeader, signerHeader) => {
		const outcome = {
			signerIndex: index,
			verified: false,
			error
		};
		if (protectedHeader) outcome.protectedHeader = protectedHeader;
		if (signerHeader) outcome.signerHeader = signerHeader;
		return outcome;
	};
	if (typeof wireSig.signature !== "string") return fail(new JWTError(`Signature[${index}] is missing the "signature" field.`, "ERR_JWS_INVALID"));
	let protectedHeader;
	try {
		protectedHeader = decodeProtectedHeader(wireSig.protected, "JWS");
	} catch (err) {
		return fail(_asJWTError(err, "ERR_JWS_INVALID"));
	}
	if (!protectedHeader.alg) return fail(new JWTError(`Invalid JWS: Signature[${index}] protected header must have an "alg" property.`, "ERR_JWS_INVALID"), protectedHeader);
	const unprotectedHeader = sanitizeObjectCopy(wireSig.header);
	try {
		_assertDisjointHeaders(protectedHeader, unprotectedHeader);
	} catch (err) {
		return fail(_asJWTError(err, "ERR_JWS_INVALID"), protectedHeader, unprotectedHeader);
	}
	const effective = _mergeHeaders(protectedHeader, unprotectedHeader);
	const alg = protectedHeader.alg;
	if (options.typ && protectedHeader.typ !== options.typ) return fail(new JWTError(`Invalid JWS: "typ" Header Parameter mismatch on signature[${index}]. Expected "${options.typ}", got "${protectedHeader.typ}".`, "ERR_JWS_INVALID"), protectedHeader, unprotectedHeader);
	let rawKeyMaterial;
	try {
		rawKeyMaterial = await keyResolver(effective, wireSig.protected ?? "");
	} catch (err) {
		return fail(_asJWTError(err, "ERR_JWK_KEY_NOT_FOUND", `Key resolver failed for signature[${index}]`), protectedHeader, unprotectedHeader);
	}
	const algError = checkAlgAllowed(alg, rawKeyMaterial, options.algorithms, JWS_ALG_CTX);
	if (algError) return fail(algError, protectedHeader, unprotectedHeader);
	const signatureBytes = base64UrlDecode(wireSig.signature, { returnAs: "uint8array" });
	const signingInputBytes = textEncoder.encode(`${wireSig.protected ?? ""}.${payload}`);
	let verified;
	try {
		verified = await _verifyWithKey(alg, rawKeyMaterial, signatureBytes, signingInputBytes, effective.kid);
	} catch (err) {
		return fail(_asJWTError(err, "ERR_JWS_SIGNATURE_INVALID"), protectedHeader, unprotectedHeader);
	}
	if (!verified) return fail(new JWTError(`JWS signature[${index}] verification failed.`, "ERR_JWS_SIGNATURE_INVALID"), protectedHeader, unprotectedHeader);
	let decodedPayload;
	try {
		decodedPayload = decodeJWSPayload(payload, protectedHeader.b64 !== false, options.forceUint8Array);
	} catch (err) {
		return fail(_asJWTError(err, "ERR_JWS_INVALID"), protectedHeader, unprotectedHeader);
	}
	try {
		validateCriticalHeadersJWS(protectedHeader, options.recognizedHeaders);
		validateJwtClaimsIfJsonPayload(decodedPayload, options);
	} catch (err) {
		return fail(_asJWTError(err, "ERR_JWT_CLAIM_INVALID"), protectedHeader, unprotectedHeader);
	}
	const outcome = {
		signerIndex: index,
		verified: true,
		payload: decodedPayload,
		protectedHeader
	};
	if (unprotectedHeader) outcome.signerHeader = unprotectedHeader;
	return outcome;
}
function _asJWTError(err, fallbackCode, fallbackMessage) {
	if (err instanceof JWTError) return err;
	return new JWTError(fallbackMessage ?? (err instanceof Error ? err.message : typeof err === "string" ? err : "Unknown error"), fallbackCode, err);
}
async function sign(payload, key, options = {}) {
	const { protectedHeader: userHeader } = options;
	let { alg } = options;
	if (!alg) if (isJWK(key) && key.alg) alg = key.alg;
	else throw new JWTError("JWS \"alg\" (Algorithm) must be provided in options", "ERR_JWS_ALG_MISSING");
	if (alg === "none") throw new JWTError("\"none\" is not a valid signing algorithm", "ERR_JWS_ALG_NOT_ALLOWED");
	const signingKey = await resolveSigningKey(alg, await importKey(key, {
		alg,
		expect: "private"
	}), "sign");
	const protectedHeader = _buildJWSHeader(alg, key, userHeader, payload);
	const computedPayload = computeJwtTimeClaims(payload, options);
	const protectedHeaderEncoded = base64UrlEncode(JSON.stringify(protectedHeader));
	let payloadBytes;
	if (payload instanceof Uint8Array) payloadBytes = payload;
	else if (typeof payload === "string") payloadBytes = textEncoder.encode(payload);
	else if (typeof payload === "object" && payload !== null) payloadBytes = textEncoder.encode(JSON.stringify(computedPayload || payload));
	else throw new TypeError("Payload must be a string, Uint8Array, or a JSON-serializable object.");
	const signingInputString = `${protectedHeaderEncoded}.${protectedHeader.b64 !== false ? base64UrlEncode(payloadBytes) : textDecoder.decode(payloadBytes)}`;
	const signingInputBytes = textEncoder.encode(signingInputString);
	return `${signingInputString}.${base64UrlEncode(await sign$1(alg, signingKey, signingInputBytes))}`;
}
async function verify(jws, key, options = {}) {
	const parts = jws.split(".");
	if (parts.length !== 3) throw new JWTError("Invalid JWS: Must contain three parts separated by dots.", "ERR_JWS_INVALID");
	const [protectedHeaderEncoded, payloadEncoded, signatureEncoded] = parts;
	const protectedHeader = decodeProtectedHeader(protectedHeaderEncoded, "JWS");
	if (!protectedHeader.alg) throw new JWTError("Invalid JWS: Protected header must be an object with an \"alg\" property.", "ERR_JWS_INVALID");
	const alg = protectedHeader.alg;
	if (options.typ && protectedHeader.typ !== options.typ) throw new JWTError(`Invalid JWS: "typ" (Type) Header Parameter mismatch. Expected "${options.typ}", got "${protectedHeader.typ}".`, "ERR_JWS_INVALID");
	let signatureBytes;
	try {
		signatureBytes = base64UrlDecode(signatureEncoded, { returnAs: "uint8array" });
	} catch {
		throw new JWTError("Invalid JWS: Signature could not be decoded.", "ERR_JWS_INVALID");
	}
	const keyInput = typeof key === "function" ? await key(protectedHeader, jws) : key;
	const algError = checkAlgAllowed(alg, keyInput, options.algorithms, JWS_ALG_CTX);
	if (algError) throw algError;
	const signingInputBytes = textEncoder.encode(`${protectedHeaderEncoded}.${payloadEncoded}`);
	if (isJWKSet(keyInput)) {
		const candidates = getJWKsFromSet(keyInput, buildJWKSetFilter(protectedHeader));
		if (candidates.length === 0) throw new JWTError(`No key found in JWK Set${protectedHeader.kid ? ` with kid "${protectedHeader.kid}"` : ""}.`, "ERR_JWK_KEY_NOT_FOUND");
		let verified = false;
		for (const candidate of candidates) if (await verify$1(alg, await resolveSigningKey(alg, await importKey(candidate, {
			alg,
			expect: "public"
		}), "verify"), signatureBytes, signingInputBytes)) {
			verified = true;
			break;
		}
		if (!verified) throw new JWTError("JWS signature verification failed.", "ERR_JWS_SIGNATURE_INVALID");
	} else if (!await verify$1(alg, await resolveSigningKey(alg, await importKey(keyInput, {
		alg,
		expect: "public"
	}), "verify"), signatureBytes, signingInputBytes)) throw new JWTError("JWS signature verification failed.", "ERR_JWS_SIGNATURE_INVALID");
	const useB64 = protectedHeader.b64 !== false;
	let payload;
	try {
		payload = decodeJWSPayload(payloadEncoded, useB64, options.forceUint8Array);
	} catch (error_) {
		throw new JWTError(`Invalid JWS: Payload decoding failed (${error_ instanceof Error ? error_.message : error_})`, "ERR_JWS_INVALID");
	}
	validateCriticalHeadersJWS(protectedHeader, options.recognizedHeaders);
	validateJwtClaimsIfJsonPayload(payload, options);
	return {
		payload,
		protectedHeader
	};
}
function _buildJWSHeader(alg, key, userHeader, payload) {
	const safeHeader = sanitizeObjectCopy(userHeader);
	const header = {};
	if (isJWK(key) && key.kid) header.kid = key.kid;
	if (safeHeader) Object.assign(header, safeHeader);
	header.alg = alg;
	applyTypCtyDefaults(header, payload);
	if (header.b64 === true) delete header.b64;
	return header;
}
export { verifyMulti as a, signMulti as i, verify as n, verifyMultiAll as o, generalToFlattenedJWS as r, sign as t };
