// unjwt@0.7.1/dist/_chunks/libs/unsecure.mjs
function safeJsonParse(json) {
	return JSON.parse(json, _jsonReviver);
}
function sanitizeObject(obj) {
	if (!obj || typeof obj !== "object") return obj;
	const seen = /* @__PURE__ */ new WeakSet();
	seen.add(obj);
	_sanitizeInPlace(obj, seen);
	return obj;
}
function sanitizeObjectCopy(obj) {
	if (!obj || typeof obj !== "object") return obj;
	return _sanitizeCopy(obj, /* @__PURE__ */ new WeakMap());
}
const _jsonReviver = (key, value) => _isDangerousKey(key) ? void 0 : value;
function _isDangerousKey(key) {
	return key === "__proto__" || key === "prototype" || key === "constructor";
}
function _sanitizeInPlace(current, seen) {
	if (Array.isArray(current)) {
		for (let i = 0; i < current.length; i++) {
			const v = current[i];
			if (v !== null && typeof v === "object" && !seen.has(v)) {
				seen.add(v);
				_sanitizeInPlace(v, seen);
			}
		}
		return;
	}
	const keys = Object.keys(current);
	for (let i = 0; i < keys.length; i++) {
		const key = keys[i];
		if (_isDangerousKey(key)) {
			delete current[key];
			continue;
		}
		const v = current[key];
		if (v !== null && typeof v === "object" && !seen.has(v)) {
			seen.add(v);
			_sanitizeInPlace(v, seen);
		}
	}
}
function _sanitizeCopy(current, seen) {
	const existing = seen.get(current);
	if (existing !== void 0) return existing;
	if (Array.isArray(current)) {
		const out = [];
		seen.set(current, out);
		for (let i = 0; i < current.length; i++) {
			const v = current[i];
			out.push(v !== null && typeof v === "object" ? _sanitizeCopy(v, seen) : v);
		}
		return out;
	}
	const out = {};
	seen.set(current, out);
	const keys = Object.keys(current);
	for (let i = 0; i < keys.length; i++) {
		const key = keys[i];
		if (_isDangerousKey(key)) continue;
		const v = current[key];
		out[key] = v !== null && typeof v === "object" ? _sanitizeCopy(v, seen) : v;
	}
	return out;
}
const _Buffer = /* @__PURE__ */ (() => {
	try {
		return globalThis.Buffer;
	} catch {
		return;
	}
})();
const _hasBuffer = typeof _Buffer?.from === "function";
function _toBuffer(data) {
	return data instanceof Uint8Array ? _Buffer.from(data.buffer, data.byteOffset, data.byteLength) : _Buffer.from(data);
}
const _hasToBase64 = !_hasBuffer && typeof Uint8Array.prototype.toBase64 === "function";
const _hasFromBase64 = !_hasBuffer && typeof Uint8Array.fromBase64 === "function";
const _b64UrlEncOpts = /* @__PURE__ */ Object.freeze({
	alphabet: "base64url",
	omitPadding: true
});
const _b64UrlDecOpts = /* @__PURE__ */ Object.freeze({ alphabet: "base64url" });
function _bytesToBase64Fallback(bytes) {
	let binary = "";
	for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
	return btoa(binary);
}
function base64Encode(data) {
	if (data == null) throw new TypeError("base64Encode: data must be a string or Uint8Array.");
	if (_hasBuffer) return _toBuffer(data).toString("base64");
	const encodedData = data instanceof Uint8Array ? data : textEncoder.encode(data);
	if (_hasToBase64) return encodedData.toBase64();
	return _bytesToBase64Fallback(encodedData);
}
function base64UrlEncode(data) {
	if (data == null) throw new TypeError("base64UrlEncode: data must be a string or Uint8Array.");
	if (_hasBuffer) return _toBuffer(data).toString("base64url");
	const encodedData = data instanceof Uint8Array ? data : textEncoder.encode(data);
	if (_hasToBase64) return encodedData.toBase64(_b64UrlEncOpts);
	return _bytesToBase64Fallback(encodedData).replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
}
function base64Decode(data, options) {
	if (data == null) throw new TypeError("base64Decode: data must be a string or Uint8Array.");
	const isBufferInput = data instanceof Uint8Array;
	const str = isBufferInput ? textDecoder.decode(data) : data;
	const decodeToString = (options?.returnAs ?? (isBufferInput ? "uint8array" : "string")) === "string";
	if (!str) return decodeToString ? "" : new Uint8Array(0);
	if (_hasBuffer) {
		const buf = _Buffer.from(str, "base64");
		return decodeToString ? buf.toString("utf8") : new Uint8Array(buf);
	}
	const decoded = _hasFromBase64 ? Uint8Array.fromBase64(str) : Uint8Array.from(atob(str), (b) => b.codePointAt(0));
	return decodeToString ? textDecoder.decode(decoded) : decoded;
}
function base64UrlDecode(data, options) {
	if (data == null) throw new TypeError("base64UrlDecode: data must be a string or Uint8Array.");
	const isBufferInput = data instanceof Uint8Array;
	let str = isBufferInput ? textDecoder.decode(data) : data;
	const decodeToString = (options?.returnAs ?? (isBufferInput ? "uint8array" : "string")) === "string";
	if (!str) return decodeToString ? "" : new Uint8Array(0);
	if (_hasBuffer) {
		const buf = _Buffer.from(str, "base64url");
		return decodeToString ? buf.toString("utf8") : new Uint8Array(buf);
	}
	let decoded;
	if (_hasFromBase64) decoded = Uint8Array.fromBase64(str, _b64UrlDecOpts);
	else {
		str = str.replace(/-/g, "+").replace(/_/g, "/");
		while (str.length % 4) str += "=";
		decoded = Uint8Array.from(atob(str), (b) => b.codePointAt(0));
	}
	return decodeToString ? textDecoder.decode(decoded) : decoded;
}
!_hasBuffer && Uint8Array.prototype.toHex;
!_hasBuffer && Uint8Array.fromHex;
const textEncoder = /* @__PURE__ */ new TextEncoder();
const textDecoder = /* @__PURE__ */ new TextDecoder();
function secureRandomBytes(length) {
	if (!Number.isInteger(length) || length < 0) throw new RangeError("length must be a non-negative integer.");
	const bytes = new Uint8Array(length);
	if (length > 0) for (let offset = 0; offset < length; offset += 65536) {
		const chunk = bytes.subarray(offset, Math.min(offset + 65536, length));
		crypto.getRandomValues(chunk);
	}
	return bytes;
}
export { base64UrlEncode as a, safeJsonParse as c, base64UrlDecode as i, sanitizeObject as l, base64Decode as n, textDecoder as o, base64Encode as r, textEncoder as s, secureRandomBytes as t, sanitizeObjectCopy as u };
